backend/controller/artefacts/oci_registry.go

```diff
@@ -0,0 +1,242 @@
+package artefacts
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/opencontainers/go-digest"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content/memory"
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras-go/v2/registry/remote/auth"
+	"oras.land/oras-go/v2/registry/remote/retry"
+
+	"github.com/TBD54566975/ftl/backend/controller/artefacts/internal/sql"
+	"github.com/TBD54566975/ftl/backend/libdal"
+	"github.com/TBD54566975/ftl/internal/model"
+	"github.com/TBD54566975/ftl/internal/sha256"
+)
+
+const (
+	ModuleArtifactPrefix = "ftl/modules/"
+)
+
+type ContainerConfig struct {
+	Registry       string
+	Username       string
+	Password       string
+	AllowPlainHTTP bool
+}
+
+type ContainerService struct {
+	host                  string
+	repoConnectionBuilder func(container string) (*remote.Repository, error)
+
+	// in the interim releases and artefacts will continue to be linked via the `deployment_artefacts` table
+	Handle *libdal.Handle[ContainerService]
+	db     sql.Querier
+}
+
+type ArtefactRepository struct {
+	ModuleDigest     sha256.SHA256
+	MediaType        string
+	ArtefactType     string
+	RepositoryDigest digest.Digest
+	Size             int64
+}
+
+func NewContainerService(c ContainerConfig, conn libdal.Connection) *ContainerService {
+	// Connect the registry targeting the specified container
+	repoConnectionBuilder := func(path string) (*remote.Repository, error) {
+		ref := fmt.Sprintf("%s/%s", c.Registry, path)
+		reg, err := remote.NewRepository(ref)
+		if err != nil {
+			return nil, fmt.Errorf("unable to connect to container registry '%s': %w", ref, err)
+		}
+
+		reg.Client = &auth.Client{
+			Client: retry.DefaultClient,
+			Cache:  auth.NewCache(),
+			Credential: auth.StaticCredential(c.Registry, auth.Credential{
+				Username: c.Username,
+				Password: c.Password,
+			}),
+		}
+		reg.PlainHTTP = c.AllowPlainHTTP
+
+		return reg, nil
+	}
+
+	return &ContainerService{
+		host:                  c.Registry,
+		repoConnectionBuilder: repoConnectionBuilder,
+		Handle: libdal.New(conn, func(h *libdal.Handle[ContainerService]) *ContainerService {
+			return &ContainerService{
+				host:                  c.Registry,
+				repoConnectionBuilder: repoConnectionBuilder,
+				Handle:                h,
+				db:                    sql.New(h.Connection),
+			}
+		}),
+	}
+}
+
+func (s *ContainerService) GetDigestsKeys(ctx context.Context, digests []sha256.SHA256) (keys []ArtefactKey, missing []sha256.SHA256, err error) {
+	set := make(map[sha256.SHA256]bool)
+	for _, d := range digests {
+		set[d] = true
+	}
+	modules, err := s.DiscoverModuleArtefacts(ctx)
```

Comment:
This should be explicitly checking the registry for each digest, not listing the full registry then iterating over the list. Ideally Oras should be sending a `HEAD` request to `/v2/<name>/blobs/<digest>` for each artifact to check.

Even then I am not sure if this is the right approach, we can still store in the database which artifact we know about, as for JVM applications even the HEAD approach will likely be slower than we would want.

We could still potentially use the HEAD approach if we don't know about the artifact, as it would allow us to re-use jars from other clusters.

Reply:
Great call and thanks for catching it at this stage. Some rethinking is due here, this ticket will track that.
https://github.com/TBD54566975/ftl/issues/3076